src/test/java/picard/sam/RevertSamTest.java

```diff
@@ -592,4 +604,135 @@ public void testHardClipRoundTrip() throws Exception {
             }
         }
     }
+
+
+    private final static boolean OUTPUT_BY_READ_GROUP = true;
+    // LOCAL_OUTPUT_MAP means the output map file itself is stored locally. The output SAM files will always be in the cloud.
+    private enum OutputMapOptions { LOCAL_OUTPUT_MAP, CLOUD_OUTPUT_MAP, NO_OUTPUT_MAP }
+
+    @DataProvider(name = "cloudTestData")
+    public Object[][] getCloudTestData() {
+        final PicardHtsPath tempCramOutputPath = PicardBucketUtils.getTempFilePath(GCloudTestUtils.TEST_OUTPUT_DEFAULT_STR, "RevertSam", ".cram");
+        return new Object[][]{
+                // Output by read group without the output map, write output bams in the cloud
+                // Since the output file names are determined within RevertSam, we can't designate the output files to be temp files.
+                // So we mark these files to be deleted. In order to avoid collisions, there cannot be another test case with the same BAM and output options.
+                {NA12878_MEDIUM, DEFAULT_CLOUD_TEST_OUTPUT_DIR, OUTPUT_BY_READ_GROUP, OutputMapOptions.NO_OUTPUT_MAP },
+                // Output by read group using the output map, write output bams in the cloud
+                {NA12878_MEDIUM, null, OUTPUT_BY_READ_GROUP, OutputMapOptions.CLOUD_OUTPUT_MAP },
+                // Output by read group using the local output map, write output bams in the cloud
+                {NA12878_MEDIUM, null, OUTPUT_BY_READ_GROUP, OutputMapOptions.LOCAL_OUTPUT_MAP },
+                // Cram input, create a cram for each read group
+                {NA12878_MEDIUM_CRAM, null, OUTPUT_BY_READ_GROUP, OutputMapOptions.CLOUD_OUTPUT_MAP },
+                // Cram input, single output cram
+                {NA12878_MEDIUM_CRAM, tempCramOutputPath, !OUTPUT_BY_READ_GROUP, OutputMapOptions.NO_OUTPUT_MAP }};
+    }
+
+    /**
+     *
+     * Write the output map for Revert Sam based on the read groups present in the header file of the given SAM file.
+     *
+     * @param sam        The read names in the header of this sam file will be used as rows
+     * @param outputBase The directory where the per read group output files will live
+     * @return The PicardHtsPath to the tsv with each row a pair (read name, output), where output points to a temp file
+o     * created in the directory specified in outputBase.
+     */
+    private PicardHtsPath getTestReadGroupMapFile(final PicardHtsPath sam, final String outputBase) {
+        final String samExtension = sam.isBam() ? ".bam" : ".cram";
+        final PicardHtsPath result = PicardBucketUtils.getTempFilePath(DEFAULT_CLOUD_TEST_OUTPUT_DIR.getURIString(), ".txt");
```

Comment:
Having said that, it is super confusing that this method has an outputBase arg, but the master file is always in the cloud, even when the outputBase is not. Is that intentional ? If so I would rename the method to something that indicates that it is cloud-specific. More likely though, the intention is that it should create the .txt file in the output base ?

Reply:
There are two paths here. One is the ".txt" file (really tsv), where each row is the read group, and the associated path where that read group SAM is written. This file is a temp file, always living in the cloud (by default)

The outputBase arg is the base path for each of the SAM files listed in the .txt file above. This can be cloud or local. 

I renamed the variables to make this a little more clear.